### DISCUSSION TITLE
Performance impact in domain logic from v4 to v5

### DISCUSSION THREAD

**Discussion by Jasenkoo** [mikro-orm/mikro-orm#3019] [Q&A]
**Describe the bug**
I don't know what is happening exactly and can't show the full code, but will try to describe it the best I can.

To be clear, noting in the logic is changed, only mikro-orm library is updated from the v4.5.10 to 5.1.2. To add, I am using `NestJS` for node framework.

This issue is reproducible almost everywhere in the app, however, it is most noticeable in one of the microservices where it handles large datasets.

General  configuration in the microservice

**main.ts**
```ts
(async (): Promise<void> => {
    const app = await NestFactory.createMicroservice<MicroserviceOptions>(AppModule, {
        transport: Transport.TCP,
        options: {
            port: +process.env.MS_SHARED_SERVICE,
        },
    });

    await app.listen();
})();
```

**app.module.ts**
```ts
@Module({
    imports: [
        MikroOrmModule.forRoot({
            entities: [ContactEntity, CompanyEntity],
            namingStrategy: EntityCaseNamingStrategy,
            type: 'postgresql',
            dbName: process.env.DB_NAME,
            host: process.env.DB_HOST,
            port: +process.env.DB_PORT,
            user: process.env.DB_USERNAME,
            password: process.env.DB_PASSWORD,
            metadataProvider: TsMorphMetadataProvider,
            allowGlobalContext: true, //  <- only diff between v4 and v5
            registerRequestContext: false,
        }),
    ],
})
export class AppModule implements NestModule {
    configure(consumer: MiddlewareConsumer): void {
        consumer.apply(MikroOrmMiddleware).forRoutes('*');
}
    
```

**app.controller.ts**

```ts
 @MessagePattern('handlePCMImport')
  async handlePCMImport(contacts: IPCMSyncData): Promise<IRLogInfo> {
      return this.appService.pcmSync(contacts);
  }
```

This was the configuration, noting big, one controller end point

Now in the service, the method is called

```ts

public async pcmSync({ fetched, existing }: IPCMSyncData): Promise<IRLogInfo> {
      // This method is doing simple persist and some additional minor reads and ends with this.orm.em.flush()

       // Then it is calling another method in the same class, as a final step
       await this.linkContactsToCompaniesAndDepartments()
}

public async linkContactsToCompaniesAndDepartments(): Promise<void> {
        const [allCompaniesAndDepartments, allContacts] = await Promise.all([
            this.orm.em.find(CompanyEntity, { deletedAt: { $eq: null } }),
            this.orm.em.find(ContactEntity, { deletedAt: { $eq: null } }),
        ]);
        // This is the format function that will do a lot of data mapping, like few thousands items in various loops
        const companiesData = categorizeContactsForCompanies(allContacts, allCompaniesAndDepartments);

        await this.orm.em.persistAndFlush(companiesData);
    }
```

When it comes to this line

```ts
const companiesData = categorizeContactsForCompanies(allContacts, allCompaniesAndDepartments);
```
it just takes too long to finish. Measured with `process.hrtime`, on v4 this takes around 20 seconds to finish, however, that same logic with v5 takes over 200 seconds to finish. To add, this method has nothing with the database operations, no loading, nothing external called, just pure data mapping.

To give you some perspective on the function

```ts
export const categorizeContactsForCompanies = (
    fetched: ContactEntity[],
    companies: CompanyEntity[],
): ContactEntity[]  {

const topLevelSupervisors = fetched
        .filter(
            (contact) =>
                contact.xkey === contact.supervisorXkey ||
                !fetched.some((superior) => superior.xkey === contact.supervisorXkey) ||
                fetched.some(
                    (superior) => superior.xkey === contact.supervisorXkey && superior.company !== contact.company,
                ),
        )

}
```

As you can see this simple filter function, does barely anything, with mikro-orm v4 takes a few seconds to finish, but when I bump orm to v5, it takes over 100 seconds to finish.

What could be causing this or am I missing something in the configuration? To add again, nothing is changed, just the update from 4 to 5, same database, same models, same logic.

**Expected behavior**
Should not affect performance in the code that is executed after the database read

**Versions**

| Dependency | Version |
| - | - |
| node | 14.17.6 |
| typescript | 4.6.3 |
| mikro-orm | 5.1.2 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
> allowGlobalContext: true, //  <- only diff between v4 and v5

Well, you definitely don't want that. Using global context basically equals to turning your app into one huge memory leak, not talking about other problems like instable serialized response. Growing identity map can definitely lead into perf issues, although it would probably behave slowly after some time, not on the first request. Even if the `RequestContext` helper dont work for your use case, you can't just use one global context and expect things to work properly, you need forks.

Anyway, can't help much with perf issues without clear reproduction, hence moving this to discussions.

**Comment by Jasenkoo** [comment]
Ok, tnx for the answer, will check with the repositories and turning off the `globalContext` and see.

**Reply by B4nan** [reply]
It sounds quite weird that the `categorizeContactsForCompanies` method would cause this, but if its like that, it should be rather easy to reproduce, as all you need is to prepare the same inputs for it artificially. I guess the real implementation is more complex, given your example is not doing anything, just filtering (but not even returning what was filtered).

**Reply by Jasenkoo** [reply]
Will make a reproduction and post it here

**Comment by Jasenkoo** [comment]
I will close this discussion, while I was creating a reproduction repository, it was working fine, not exactly the same code, but close enough. I will see to disable `globalContext` and use `.fork` where needed. But it looks like it is something in the app.

**Comment by Jasenkoo** [comment]
@B4nan After some further research, discovered the following:

What I did:

1.  Removed `allowGlobalContext`
2.  Added `@UseRequestContext` decorator
3. Removed all other db operations 

In the end, the only code that was left was this:

```ts
const em = this.orm.em.fork();

const [allCompaniesAndDepartments, allContacts] = await Promise.all([
   em.find(CompanyEntity, { deletedAt: { $eq: null } }),
   em.find(ContactEntity, { deletedAt: { $eq: null } }),
]);

const companiesData = categorizeContactsForCompanies(allContacts, allCompaniesAndDepartments);
```

As I mentioned 

> This issue is reproducible almost everywhere in the app, however, it is most noticeable in one of the microservices where it handles large datasets.

The issue can be noticed where some larger amount of data is handled.

`ContactEntity` is an entity with 34 fields. It only has 2 relations to the `CompanyEntity` like this:

```ts
@Entity({ tableName: 'contacts', customRepository: () => ContactRepository })
@Filter({ name: 'contact_paranoid', cond: { deletedAt: null }, default: true })
export class ContactEnity {
   // Rest of the fields are just varchar(255) or int
  @ManyToOne(() => CompanyEntity, {
    joinColumn: 'contactDepartmentId',
    name: 'contactDepartmentId',
    cascade: [Cascade.PERSIST, Cascade.MERGE],
    nullable: true,
  })
  contactDepartment: CompanyEntity;

  @ManyToOne(() => CompanyEntity, {
        joinColumn: 'contactCompanyId',
        name: 'contactCompanyId',
        cascade: [Cascade.PERSIST, Cascade.MERGE],
        nullable: true,
   })
  contactCompany: CompanyEntity;
}
```

`CompanyEntity` is quite smaller, with only 9 columns and has 2 self referencing relations and one to contacts like this

```ts
@Entity({ tableName: 'companies', customRepository: () => CompanyRepository })
@Filter({ name: 'company_paranoid', cond: { deletedAt: null }, default: true })
export class CompanyEntity {
    @ManyToOne(() => CompanyEntity , { joinColumn: 'parentId', nullable: true })
    parent: CompanyEntity;

    @OneToMany(() => CompanyEntity , (company) => company.parent, {
        joinColumn: 'parentId',
        fieldName: 'id',
        inverseJoinColumn: 'children',
    })
    children = new Collection<CompanyEntity >(this);

    @OneToMany(() => ContactEnity, (contact) => contact.contactDepartment)
    contacts = new Collection<ContactEnity>(this);
}
```

As you can see, this company entity is connected for a tree structure, thus, `categorizeContactsForCompanies ` is used to categorize contacts into proper relations to companies.

Now, going back to this code

```ts
const em = this.orm.em.fork();

const [allCompaniesAndDepartments, allContacts] = await Promise.all([
   em.find(CompanyEntity, { deletedAt: { $eq: null } }),
   em.find(ContactEntity, { deletedAt: { $eq: null } }),
]);

const companiesData = categorizeContactsForCompanies(allContacts, allCompaniesAndDepartments);
```

`em.find(ContactEntity, { deletedAt: { $eq: null } })` will return approximate around 10000 contacts, sometimes more, sometimes less.

However, as mentioned that this line 

```ts
const companiesData = categorizeContactsForCompanies(allContacts, allCompaniesAndDepartments);
```
takes way longer in v5 than v4 and what it does is a simple format and further down the line some mapping.

But when I do  parse on the read data to get plain JSON like:

```ts
const companiesData = categorizeContactsForCompanies(
   JSON.parse(JSON.stringify(allContacts)),
   JSON.parse(JSON.stringify(allCompaniesAndDepartments)),
);
```

It finishes much faster, like in v4. It looks like the Entity class is causing it to be much slower, I don't know what else could be done here? Any feedback would be appriciated?

**Reply by Jasenkoo** [reply]
To add, my reproduction repo didn't include a more complex data structure, just plain columns, so the issue might be related to relationships, that's why it wasn't noticeable.

**Reply by B4nan** [reply]
In v5 we track changes done on the entity so we can autoflush when needed - all properties are redefined on the prototype to allow this, so we know when an entity is changed/touched. In v4 only relation properties were redefined this way (so we could implement propagation), in v5 we do it for scalar properties too. It's weird that it would cause such perf hit, as for scalar properties it is rather simple get/set pair that works with a hidden `data` object.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityHelper.ts#L112-L129

**Reply by B4nan** [reply]
Note that for managed entities, this happens elsewhere - we cache the property definitions in `EntityMetadata` class (so create the definition once during discovery phase, once per entity type) and set them directly on the entity instance in `EntityFactory`. So for managed entities, we don't use this "double setter" trick as in `EntityHelper`. 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/typings.ts#L362-L383
https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityFactory.ts#L202

So for managed entities, it should be all set by the time you work with the entities.

**Reply by Jasenkoo** [reply]
Hmm, strange. Also when I mentioned I noticed somewhere else. There is one pagination method in the app, the entity is quite complex, with many relations, `OneToMany`, `ManyToMany` . There is also quite a large dataset, where one item can have a few thousands in `@OneToMany`. The read method is done very fast, a few ms, since it only reads 30 items. However, after that read, we do another `Promise.all()` to count some additional data and map to each item. The difference between v4 and v5 is also quite noticeable too.

**Reply by Jasenkoo** [reply]
I will revert to v4 for now, I am out of options. Tried some different configs, and even moved the logic from microservice just in case, but without luck. On the main server where there are a lot of entities, it is even slower than it is in ms since it is only 2 entities in the service and over 100 on the main server.

Since it is a `NestJS` microservice,  it is easy to convert to the app server, tried that without starting anything else, same issue. I will copy the exact code and create a repo so you can check

**Reply by B4nan** [reply]
Ok thanks, I just tried a simple test on 10k entities, where I change one scalar 1000 times for each entity (so 10 mil property changes), it is also significantly slower than with POJOs (~3 seconds vs ~1,2s). It is definitely caused by the get/set pairs, if I comment those out, its pretty much the same as with POJO. Will try to experiment with that first, worst case I can offer a way to disable the change tracking on scalar properties.

**Reply by Jasenkoo** [reply]
@B4nan [Repository](https://github.com/Jasenkoo/mikro-orm-v4-v5-performace) created. You will see that there is quite a bit of performance difference between version

**Reply by B4nan** [reply]
Did some initial experiments, a simple get/set property is still quite slow (and we use that for many things), but we might be able to use `Proxy` instead, looks like it's much faster. And with `Proxy` we could also implement delete traps, something that was not possible with the current approach (e.g. to propagate `delete book.author` into `author.books` collection, now it works only if you explicitly set `null/undefined`).

**Reply by Jasenkoo** [reply]
Cool, if it can be achieved with Proxies. I don't have much experience with them, but I know Vue uses them for reactive properties. Sry I can't be more helpful since I am not familiar much with the mikro-orm code itself, maybe when I have time, I take a deeper look. Anyways, a very nice library and the DX is on point 🥇

**Reply by B4nan** [reply]
~Btw do you have `useDefineForClassFields` enabled in tsconfig? Looks like it's fast if you disable that (it is true when targetting recent node versions, e.g. `es2020`). (testing just my small repro)~ actually not, a mismeasurement on my side

**Reply by B4nan** [reply]
Looks like this is more complicated, in other perf tests the solution with proxies is much worse (in my bechmark its takes 50% more time with proxies). I will take a look at your repro, maybe what I got it too artificial.

edit: looks like proxies dont help with your repro either

**Comment by B4nan** [comment]
FYI I've been looking into this several times already, and a workaround I found is to first map the entities to the entity data directly, its similar to doing the serialization but should be more performant:

```ts
export const categorizeContactsForCompanies = (fetched: Contact[], companies: Company[]): Contact[] => {
    const allCompanies = companies;
    const entityData = fetched.map((c) => wrap(c, true).__data);
    const topLevelSupervisors = fetched
        .filter(
            (contact, i) =>
                entityData[i].xkey === entityData[i].supervisorXkey ||
                !fetched.some((superior, j) => entityData[j].xkey === entityData[i].supervisorXkey) ||
                fetched.some(
                    (superior, k) => entityData[k].xkey === entityData[i].supervisorXkey && entityData[k].company !== entityData[i].company,
                ),
        )
        .filter((supervisor, i) =>
            allCompanies
                .filter((company, l) => !entityData[l].isDepartment)
                .some((company, l) => entityData[l].name === entityData[i].company),
        );

    return fetched;
};
```

Trying to find a way to allow disabling the change tracking on property level, it is used only for auto flush detection so it should be feasible for fast property access. You can always explicitly persist the entity even when its managed and it will serve the same purpose, so there is a workaround even when the detection would be disabled.

edit: [this](https://github.com/mikro-orm/mikro-orm/commit/7d5e32d8eb7ec25ace2336e1bd4da5d97d015c08) should help, but its still way slower than the workaround above (~10ms vs ~2500ms), god knows why, haven't checked this in detail yet...
